File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/TextSizer.java
package com.brentpanther.bitcoinwidget;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Paint;
import android.graphics.Rect;
import android.text.TextUtils;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

abstract class TextSizer {

    static float getTextSize(Context context, String text, Pair<Integer, Integer> availableSize) {
        @SuppressLint("InflateParams")
        ViewGroup vg = (ViewGroup) LayoutInflater.from(context).inflate(R.layout.widget_layout, null);
        TextView textView = vg.findViewById(R.id.price);
        return getHighestInBounds(textView, text, availableSize.first, availableSize.second);
    }

    static float getLabelSize(Context context, String text, Pair<Integer, Integer> availableSize) {
        @SuppressLint("InflateParams")
        ViewGroup vg = (ViewGroup) LayoutInflater.from(context).inflate(R.layout.widget_layout, null);
        TextView textView = vg.findViewById(R.id.exchange);
        return getHighestInBounds(textView, text, availableSize.first, availableSize.second);
    }

    private static float getHighestInBounds(TextView textView, String text, float widthPx, float heightPx) {
        Paint paint = textView.getPaint();
        if (TextUtils.isEmpty(text)) {
            return 0;
        }
        Rect rect = new Rect();
        float dp = 6f;
        float step = 0.5f;
        while (true) {
            paint.setTextSize(dp);
            paint.getTextBounds(text, 0, text.length(), rect); // does not give accurate width
            float measuredHeight = rect.height();
            float measuredWidth = paint.measureText(text);
            if (measuredHeight > heightPx || measuredWidth >= widthPx) {
                return dp - step;
            }
            dp += step;
        }
    }

}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/TextSizer.kt
package com.brentpanther.bitcoinwidget

import android.content.Context
import android.graphics.Rect
import android.util.Pair
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import kotlin.math.roundToLong

internal object TextSizer {

    fun getTextSize(context: Context, text: String, availableSize: Pair<Int, Int>, layout: Int): Float {
        val vg = LayoutInflater.from(context).inflate(layout, null) as ViewGroup
        val textView = vg.findViewById<TextView>(R.id.price)
        return getHighestInBounds(textView, text, availableSize.first.toFloat(), availableSize.second.toFloat())
    }

    fun getLabelSize(context: Context, text: String, availableSize: Pair<Int, Int>, layout: Int): Float {
        val vg = LayoutInflater.from(context).inflate(layout, null) as ViewGroup
        val textView = vg.findViewById<TextView>(R.id.exchange)
        return getHighestInBounds(textView, text, availableSize.first.toFloat(), availableSize.second.toFloat())
    }

    private fun getHighestInBounds(textView: TextView, text: String, widthPx: Float, heightPx: Float): Float {
        val paint = textView.paint
        if (text.isEmpty()) {
            return 0f
        }
        val rect = Rect()
        var dp = 6f
        val step = 0.5f
        while (true) {
            //TODO: this isn't working right
            paint.textSize = dp
            paint.getTextBounds(text, 0, text.length, rect) // does not give accurate width
            val measuredHeight = rect.height().toFloat()
            val measuredWidth = paint.measureText(text)
            if (measuredHeight > heightPx || measuredWidth >= widthPx) {
                return (2 * (dp - step)).roundToLong() / 2.0f
            }
            dp += step
        }
    }

}
